packages/protocol/ledger/conflictdag/conflictdag.go

```diff
@@ -139,53 +190,112 @@ func (b *ConflictDAG[ConflictIDType, ConflictingResourceID]) UnconfirmedConflict
+
+		confirmationWalker.PushAll(conflict.Parents().Slice()...)
 
-			b.Utils.forEachConflictingConflictID(conflict, func(conflictingConflictID ConflictID) bool {
-				rejectionWalker.Push(conflictingConflictID)
-				return true
-			})
+		conflict.ForEachConflictingConflict(func(conflictingConflict *Conflict[ConflictIDType, ResourceIDType]) bool {
+			conflictsToReject.Add(conflictingConflict)
+			return true
 		})
 	}
 
+	modified = c.rejectConflictsWithFutureCone(conflictsToReject) || modified
+
+	// // Delete all resolved ConflictSets (don't have a pending conflict anymore).
+	// for it := conflictSets.Iterator(); it.HasNext(); {
+	//	conflictSet := it.Next()
+	//
+	//	pendingConflicts := false
+	//	for itConflict := conflictSet.Conflicts().Iterator(); itConflict.HasNext(); {
+	//		conflict := itConflict.Next()
+	//		if conflict.ConfirmationState() == confirmation.Pending {
+	//			pendingConflicts = true
+	//			continue
+	//		}
+	//		conflict.deleteConflictSet(conflictSet)
+	//	}
+	//
+	//	if !pendingConflicts {
+	//		c.conflictSets.Delete(conflictSet.ID())
+	//	}
+	// }
+	//
+	// // Delete all resolved Conflicts that are not part of any ConflictSet anymore.
+	// for it := conflicts.Iterator(); it.HasNext(); {
+	//	conflict := it.Next()
+	//	if conflict.ConflictSets().Size() == 0 {
+	//		c.conflicts.Delete(conflict.ID())
+	//	}
+	// }
+
+	return modified
+}
+
+func (c *ConflictDAG[ConflictIDType, ResourceIDType]) rejectConflictsWithFutureCone(initialConflicts *set.AdvancedSet[*Conflict[ConflictIDType, ResourceIDType]]) (modified bool) {
+	rejectionWalker := walker.New[*Conflict[ConflictIDType, ResourceIDType]]().PushAll(initialConflicts.Slice()...)
 	for rejectionWalker.HasNext() {
-		b.Storage.CachedConflict(rejectionWalker.Next()).Consume(func(conflict *Conflict[ConflictID, ConflictingResourceID]) {
-			if modified = conflict.setConfirmationState(confirmation.Rejected); !modified {
-				return
-			}
+		conflict := rejectionWalker.Next()
+		if !conflict.setConfirmationState(confirmation.Rejected) {
+			continue
+		}
 
-			b.Events.ConflictRejected.Trigger(conflict.ID())
+		modified = true
 
-			b.Storage.CachedChildConflicts(conflict.ID()).Consume(func(childConflict *ChildConflict[ConflictID]) {
-				rejectionWalker.Push(childConflict.ChildConflictID())
-			})
-		})
+		c.Events.ConflictRejected.Trigger(conflict)
+		rejectionWalker.PushAll(conflict.Children().Slice()...)
 	}
 
 	return modified
 }
 
 // ConfirmationState returns the ConfirmationState of the given ConflictIDs.
-func (b *ConflictDAG[ConflictID, ConflictingResourceID]) ConfirmationState(conflictIDs *set.AdvancedSet[ConflictID]) (confirmationState confirmation.State) {
-	b.RLock()
-	defer b.RUnlock()
+func (c *ConflictDAG[ConflictIDType, ResourceIDType]) ConfirmationState(conflictIDs *set.AdvancedSet[ConflictIDType]) (confirmationState confirmation.State) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	// TODO: simplify this method
+
+	confirmationState = confirmation.Confirmed
```

Comment:
Why defaulting to Confirmed instead of Undefined? If the conflict set received is empty you would return Confirmed.